File: src/main/java/com/baidu/algorithm/backtracking/_131_Palindrome_Partitioning.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        res = []
        self.dfs(res, [], s)

        return res

    def dfs(self, res, cur_res, s):
        if not s:
            res.append(cur_res)
        else:
            for i in xrange(len(s)):
                if self.valid_palindrome(s[:i + 1]):
                    self.dfs(res, cur_res + [s[:i + 1]], s[i + 1:])

    def valid_palindrome(self, s):
        return s == s[::-1]

File: src/main/java/com/baidu/algorithm/bfs/_131_Palindrome_Partitioning.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.bfs;

import java.util.ArrayList;
import java.util.List;

/**
 * _131_Palindrome_Partitioning
 *
 * @author xuhaoran01
 */
public class _131_Palindrome_Partitioning {

    private boolean isPalindrome(String s) {

        for (int i = 0, j = s.length() - 1; i < j; i++, j--) {
            if (s.charAt(i) != s.charAt(j)) {
                return false;
            }
        }

        return true;
    }

    private void dfs(List<List<String>> res, List<String> curRes, String s) {
        if (s.isEmpty()) {
            res.add(new ArrayList<>(curRes));
        } else {
            for (int i = 1; i <= s.length(); i++) {
                if (isPalindrome(s.substring(0, i))) {
                    curRes.add(s.substring(0, i));
                    dfs(res, curRes, s.substring(i));
                    curRes.remove(curRes.size() - 1);
                }
            }
        }
    }

    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        if (s == null || s.isEmpty()) {
            return res;
        }

        dfs(res, new ArrayList<>(), s);

        return res;
    }
}
